lib/errordedup/errordedup.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package errordedup
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// timeWindow is the time window over which errors are deduplicated.
+	timeWindow = time.Minute
+	// timeWindowCleanupInterval is the interval between cleanups of
+	// time windows that have already ended.
+	// Since the time window is set to 1 minute and the cleanup interval
+	// is set to 10 seconds, time windows can be in fact slightly larger
+	// than 1 minute (i.e., 1 minute and 10 seconds, in the worst case).
+	timeWindowCleanupInterval = 10 * time.Second
+)
+
+// Config contains the error deduplicator config.
+type Config struct {
+	// Entry is the logger entry.
+	Entry *log.Entry
+	// LogLevel is the log level at which errors will be logged.
+	// The default log level (0) is panic.
+	LogLevel log.Level
+	// DebugReport indicates whether a debug report should be generated
+	// for the errors reported.
+	DebugReport bool
+	// ErrorSubstrings contains a list of substrings belonging to the
+	// errors that should be deduplicated.
+	ErrorSubstrings []string
+	// ChannelSize is the size of the channel used to send messages
+	// to the error deduplicator.
+	ChannelSize int
+	// Clock is a clock to override in tests, set to real time clock
+	// by default.
+	Clock clockwork.Clock
+}
+
+// CheckAndSetDefaults verifies configuration and sets defaults
+func (c *Config) CheckAndSetDefaults() error {
```

Comment:
nit: Make this private, only New should be calling it.

Reply:
Done in c56f31e04b2349e1bf6666e86603e64562a8b8a6.